### ISSUE TITLE
-O0 vs -O2: -O0 results in miscompilation?

### ISSUE THREAD

**Issue by waywardmonkeys** [dylan-lang/opendylan#921]
I was trying building ``melange`` by modifying my ``config.jam`` to tell the compilation to use ``-g -O0`` instead of ``-g -O2``.

This resulted in an error at run-time that wasn't present before:

```
24 is not of type {<class>: <application-process>}
24 is not of type {class <application-process> 0x57e39c}
```

(Getting this error requires some code that I haven't landed yet.)

**Comment by waywardmonkeys** [comment]
This happened with this code:

```dylan
define function get-compiler-defines (cmd :: <string>)
  local method parse-define (line :: <string>)
          let define-start = size("#define ");
          let sep = find-substring(line, " ", start: define-start);
          let def = copy-sequence(line, start: define-start, end: sep);
          let val = copy-sequence(line, start: sep + 1);
          values(def, val)
        end;
  with-application-output (stream = cmd)
    let output = read-to-end(stream);
    let lines = split(strip(output), "\n");
    let defines = make(<stretchy-vector>);
    do(method(line)
        let (d, v) = parse-define(line);
        unless (find-substring(v, "##"))
          add!(defines, d);
          add!(defines, v);
        end;
       end, lines);
    defines
  end;
end function;
```

The ``with-application-output`` is:

```dlan
define macro with-application-output
  { with-application-output (?stream:variable = ?command:expression,
                             #rest ?keys:expression)
      ?body:body
    end }
  => { begin
         let _stream = #f;
         let _process = #f;
         block ()
           let (exit-code, signal, process, ostream)
             = run-application(?command, asynchronous?: #t, output: #"stream",
                               ?keys);
           _stream := ostream;
           _process := process;
           let ?stream :: <file-stream> = ostream;
           ?body;
         cleanup
           if (_process) wait-for-application-process(_process) end;
           if (_stream & stream-open?(_stream)) close(_stream) end;
         end
       end }
end macro with-application-output;
```

```dylan
METHOD get-compiler-include-directories (cmd :: <string>) => (#rest results)
  {{ _stream }} := make-cell(^#f)
  {{ _process }} := make-cell(^#f)
  UNWIND-PROTECT (t3)
    *t5(4) := [CONGRUENT-CALLi ^{<&method> run-application (type-union(<string>, limited(<sequence>, #"of", <string>)))}({{ cmd }}, ^#[], ^#t, ^#t, ^#t, ^#f, ^#f, ^#f, ^#t, ^#f, ^#f, ^#"inherit", ^#"signal", ^#"stream", ^#"replace", ^#"inherit", ^#"replace")]
    {{ process }} := *t5(4) [2]
    {{ ostream }} := *t5(4) [3]
    cell-value({{ _stream }}) := {{ ostream }}
    cell-value({{ _process }}) := {{ process }}
    {{ stream }} := check-type {{ ostream }} :: ^{<&class> <file-stream>}
    {{ output }} := [CONGRUENT-CALLg ^{<&generic> read-to-end}({{ stream }})]
    t18 := [CALLg ^{<&generic> strip}({{ output }})]
    *t19(0,#rest) := [CONGRUENT-CALLg ^{<&generic> split}(t18, ^":", ^#[])]
    end-protected-block entry-state: t3
  CLEANUP
    t28 := cell-value({{ _process }})
    IF (t28)
      t29 := cell-value({{ _process }})
      [CALLx ^{<&method> wait-for-application-process (<application-process>)}(t29)]
    ELSE
    END IF
    [IF-MERGE #f #f]
    t25 := cell-value({{ _stream }})
    IF (t25)
      t26 := cell-value({{ _stream }})
      t20 := [CALLg ^{<&generic> stream-open?}(t26)]
    ELSE
    END IF
    t21 := [IF-MERGE t20 ^#f]
    IF (t21)
      t24 := cell-value({{ _stream }})
      [CALLg ^{<&generic> close}(t24)]
    ELSE
    END IF
    [IF-MERGE #f #f]
    end-cleanup-block entry-state: t3
  END UNWIND-PROTECT
  return *t19(0,#rest)
END
```

```c
dylan_value Kget_compiler_include_directoriesYportabilityVmelange_coreI (dylan_value cmd_) {
  dylan_value T1;
  dylan_value T2_0;
  dylan_value T2_1;
  dylan_value T2_2;
  dylan_value T2_3;
  dylan_value process_;
  dylan_value ostream_;
  dylan_value stream_;
  dylan_value output_;
  dylan_value T7;
  dylan_value T8;
  dylan_value T9;
  dylan_value T10;
  dylan_value Ustream_;
  dylan_value T12;
  dylan_value T13;
  dylan_value T14;
  dylan_value Uprocess_;
  dylan_value T16;
  dylan_value T17;

  Ustream_ = &KPfalseVKi;
  Uprocess_ = &KPfalseVKi;
  ENTER_UNWIND_FRAME(T1);
  if (!nlx_setjmp(FRAME_DEST(T1))) {
    T2_0 = Krun_applicationYoperating_systemVsystemI(cmd_, &KPempty_vectorVKi, &KPtrueVKi, &KPtrueVKi, &KPtrueVKi, &KPfalseVKi, &KPfalseVKi, &KPfalseVKi, &KPtrueVKi, &KPfalseVKi, &KPfalseVKi, IKJinherit_, &KJsignal_, IKJstream_, &KJreplace_, IKJinherit_, &KJreplace_);
    T2_1 = MV_GET_ELT(1);
    T2_2 = MV_GET_ELT(2);
    T2_3 = MV_GET_ELT(3);
    process_ = T2_2;
    ostream_ = T2_3;
    Ustream_ = ostream_;
    Uprocess_ = process_;
    primitive_type_check(ostream_, &KLfile_streamGYstreams_internalsVio);
    stream_ = ostream_;
    CONGRUENT_CALL_PROLOG(&Kread_to_endYstreamsVio, 1);
    output_ = CONGRUENT_CALL1(stream_);
    T7 = CALL1(&KstripVstrings, output_);
    CONGRUENT_CALL_PROLOG(&KsplitYcommon_extensionsVcommon_dylan, 3);
    T8 = CONGRUENT_CALL3(T7, &K4, &KPempty_vectorVKi);
    FALL_THROUGH_UNWIND(T8);
  }
  T16 = Uprocess_;
  if (T16 != &KPfalseVKi) {
    T17 = Uprocess_;
    CALL1(&Kwait_for_application_processYoperating_systemVsystem, T17);
  }
  T13 = Ustream_;
  if (T13 != &KPfalseVKi) {
    T14 = Ustream_;
    T9 = CALL1(&Kstream_openQYstreams_protocolVcommon_dylan, T14);
    T10 = T9;
  } else {
    T10 = &KPfalseVKi;
  }
  if (T10 != &KPfalseVKi) {
    T12 = Ustream_;
    CALL1(&KcloseYcommon_extensionsVcommon_dylan, T12);
  }
  CONTINUE_UNWIND();
  return(T8);
}
```

**Comment by waywardmonkeys** [comment]
So what happened here is that the values in the generated C weren't flagged as ``volatile`` ...

I suspect that the check for whether or not something should be ``volatile`` in ``dfmc-c-back-end`` should iterate recursively over the computations?

**Comment by waywardmonkeys** [comment]
Moving into the compiler ... the code involved here to determine if a variable should be volatile or not is:

```dylan
    let volatile?
      = block (result)
          for-computations (c in o)
            if (instance?(c, <block>) & ~c.entry-state.local-entry-state?)
              result(#t);
            end if;
          end for-computations;
          #f
        end block;
```

The ``entry-state`` on the ``<unwind-protect>`` is:

```
  [entry-state] = 0x039423f0 {<entry-state>} {
    [users] = 0x039bf970 {<pair>}
    [%generator] = 0x03948cc0 {<unwind-protect>}
    [%environment] = 0x0197c5a0 {<lambda-lexical-environment>}
    [temporary-properties] = 0x00040021 {<integer>: 65544}
    [%node-id] = 0x0000002d {<integer>: 11}
    [emitted-name] = 0x000e6730 {<boolean>: #f}
    [name] = 0x000e6730 {<boolean>: #f}
    [me-block] = 0x03948cc0 {<unwind-protect>}
    [exits] = 0x000e6740 {<empty-list>: #()}
  }
```

``temporary-properties`` is a packed slot, which is defined by:

```dylan
define packed-slots temporary-properties (<temporary>, <object>)
  eval slot closed-over? = #f,
    field-size:      2,
    pack-function:   pack-false-or-field,
    unpack-function: unpack-false-or-field;
  field slot frame-offset = 0, field-size: $max-frame-offset-field-size;
  quadstate slot dynamic-extent? = #"unknown";
end packed-slots;

define leaf packed-slots temporary-properties (<entry-state>, <temporary>)
  boolean slot local-entry-state? = #f;
end packed-slots;
```

In the generated C, we can see that it was looking at the value like this:

```c
    T8 = SLOT_VALUE_INITD(T7, 3);
    T9 = primitive_cast_integer_as_raw(T8);
    T10 = primitive_machine_word_logbitQ(18,T9);
```

This is the same as:

```c
  T8 = SLOT_VALUE_INITD(T7, 3);
  T10 = RAWASBOOL((1UL << 18) & (unsigned long)T8));
```

So, we can see that bits 16 and 3 are set in this field, but no others (``65544`` is ``0b00000000000000010000000000001000``).

**Comment by waywardmonkeys** [comment]
So, if we go to look at what sets ``local-entry-state?``, we find that the only thing that sets it is this:

```dylan
define method analyze-block
    (env :: <lambda-lexical-environment>, entry :: <entry-state>)
 => (local? :: <boolean>)
  // TODO: no intervening blocks (currently overly conservative)
  let block-c :: <block> = entry.generator;
  entry.local-entry-state?
    := every?(curry(local-exit?, env), entry.exits)
    & ~computations-exit-with-intervening-block?(block-c.body,
             block-c.next-computation,
             entry, #f)
end method;
```

This is something that @housel recently modified, so we're on to something now...

**Comment by waywardmonkeys** [comment]
The change to ``analyze-block`` was in 03e8de7a01e6701ca6b78209861358906f6f09e5.

**Comment by waywardmonkeys** [comment]
@housel said on IRC:

> Looking at the DFM for this function, I suspect that analyze-block is fine, it's just
> that ``<unwind-protect>`` should always cause volatile to be used (if it contains
> any function calls at all)

And I can see that the only other uses of ``local-entry-state?`` are for handling emitting computations involving ``<bind-exit>`` and ``<exit>`` ... so I'll modify the ``volatile?`` check in the C back-end to handle this a bit better.

**Comment by waywardmonkeys** [comment]
A followup comment from @housel indicates that not checking the UWP for function calls is probably safest. It is also the easiest thing to do!

**Comment by waywardmonkeys** [comment]
Ends up that fixing the above doesn't fix the actual issue at hand ... that's apparently something else going wrong.

**Comment by waywardmonkeys** [comment]
@housel and I tracked this down to being an uninitialized memory read that was happening due to an MV-type-check-rest not setting everything up correctly.

@housel fixed this in d39243638c0eae479847a45fa6109caaa692a265.
